### PR TITLE
Merge fix_rvm

### PR DESCRIPTION
**Pull request by jcarreira**
Request to merge fix_rvm.

Main changes:
1) Support for allocations of different sizes
2) Freed blocks are actually freed from block table
3) Commit functionality
4) RVM layer does not need to know about remote addresses (just about tags) - this required some small changes to the rmem layer

All the tests seem to be passing (except for the big test commit, needs to be checked)

### CODE REVIEW

File: common.c [hunk 1]
```diff
@@ -168,7 +173,8 @@ void rc_client_loop(const char *host, const char *port, void *context)
 
     TEST_Z(ec = rdma_create_event_channel());
     TEST_NZ(rdma_create_id(ec, &conn, NULL, RDMA_PS_TCP));
-    TEST_NZ(rdma_resolve_addr(conn, NULL, addr->ai_addr, TIMEOUT_IN_MS));
+    //TEST_NZ(rdma_resolve_addr(conn, NULL, addr->ai_addr, TIMEOUT_IN_MS));
+    rdma_resolve_addr(conn, NULL, addr->ai_addr, TIMEOUT_IN_MS);
```
Comment (zhemao) [inline common.c:177]: Why was the TEST_NZ removed?

**Comment by zhemao** [comment]
I should have mentioned this earlier, but I had already implemented a list and hash table data structure for rmem.c. It seems you've re-implemented them yourself here. We might want to consider merging them.

File: common.c (hunk 1 above)
Comment (jcarreira) [inline common.c:177]: Hum, my mistake. Will revert that.

File: rvm.c [hunk 2]
```diff
@@ -141,21 +150,77 @@ rvm_txid_t rvm_txn_begin(rvm_cfg_t cfg)
...
+
+bool rvm_txn_commit(rvm_cfg_t* cfg, rvm_txid_t txid)
 {
     /* TODO This is a brain-dead initial implementation. We simply copy over
      * everything in the block table, even if it wasn't modified. The long-term
      * solution is to use protection bits on the page table to automatically
      * detect changes and only copy those.
-     *
-     * TODO This also is not atomic. The current plan is to add a simple
-     * transaction to the rmem_put API.
```
Comment (NathanTP) [inline rvm.c:152]: I think this is now atomic right? I mean, you copy the pages over to shadow pages and then use rmem_server's multi-copy interface. I think we can remove these comments.

Comment (jcarreira) [inline rvm.c:152]: Yes, the last 3 lines of the comment (about the atomicity) were removed. The first 4 lines still hold and were not changed.

File: Makefile [hunk 3]
```diff
@@ -1,24 +1,56 @@
 .PHONY: clean
 
-CFLAGS  := -Wall -g -fPIC -std=gnu11
```
Comment (NathanTP) [inline Makefile:3]: Why not use c11?

**Comment by jcarreira** [comment]
Yes, I saw there was a list and hash implementation. I didn't use them because they were mingled with other code.

Yes, that should be refactored.

**Comment by NathanTP** [comment]
Why are there problems with C11/99?

On Fri, Apr 24, 2015 at 12:10 PM, João Carreira notifications@github.com
wrote:

> Yes, I saw there was a list and hash implementation. I didn't use them
> because they were mingled with other code.
> 
> Yes, that should be refactored.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zhemao/rmem-server/pull/1#issuecomment-96038950.

## 

Nathan Pemberton

**Comment by jcarreira** [comment]
-std=c11 breaks the code in a bunch of places, but -std=gnu11 seems to work well. Will fix that.

**Comment by NathanTP** [comment]
Now that I think about it, I tried c11 first and then had to switch to gnu.
The problem is that stuff like getopt are actually gnu extensions and are
only defined in GNU's version of unistd.h.

On Fri, Apr 24, 2015 at 12:15 PM, João Carreira notifications@github.com
wrote:

> -std=c11 breaks the code in a bunch of places, but -std=gnu11 seems to
> work well. Will fix that.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zhemao/rmem-server/pull/1#issuecomment-96039776.

## 

Nathan Pemberton

File: rvm.c [hunk 4]
```diff
@@ -141,21 +150,77 @@ rvm_txid_t rvm_txn_begin(rvm_cfg_t cfg)
...
+        char* block = (char*)malloc(sizeof(char) * blk->size);
+        struct ibv_mr* block_mr = rmem_create_mr(block, sizeof(char) * blk->size);
+
+        int err = rmem_get(&(cfg->rmem), block, block_mr,
+                bx + 1, blk->size);
+        
+        RETURN_ERROR(err != 0, false,
+                ("Block %d does not match replica\n", bx));
+
+        err = memcmp(block, blk->local_addr, sizeof(char) * blk->size);
```
Comment (NathanTP) [inline rvm.c:198]: Should you be checking this err?

**Pull request merged**